Add code docs to the Analytics classes, make `get_order_statuses` non-internal

tomalec opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Add code docs to the Analytics classes and make `get_order_statuses` non-internal.

Document analytics classes and some methods. Add simple examples of how to use it. Give a brief "why?" and "when?" answers. Add links between related classes.


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Review the code docs.
1. Generate [code reference](https://github.com/woocommerce/code-reference) for the latest release: `./deploy.sh -v --build-only -s 9.1.4`
2. Checkout this branch and build woocommerce.zip: 
```sh
git checkout dev/fix-phpdoc-code-reference
cd plugins/woocommerce
pnpm run build:zip
```
4. unzip file to the code-reference directory
5. Generate again from local changes ``./deploy.sh -v --build-only -s 9.1.4 --no-download``
6. Review generated pages:
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-GenericController.html`
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-OrderAwareControllerTrait.html`
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-DataStore.html`
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-GenericStatsController.html`
 - `file:///path/to/code-reference/build/api/classes/Automattic-WooCommerce-Admin-API-Reports-StatsDataStoreTrait.html`

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

tomalec committed:
Add code docs to the Analytics classes, make `get_order_statuses` non-internal

tomalec commented:
I was not 100% sure if we should make `OrderAwareControllerTrait` internal or public. 

Previously, those methods were internal (the entire Report\Controller class was internal). And I'd like to avoid exposing too much as then backward compatibility would limit our pace of improvements. 
But I think it's worth exposing at least `get_order_statuses` to 3PDs to make reports consistent in terms of which orders are taken into account. However, that method specifically marked as internal in https://github.com/woocommerce/woocommerce-admin/pull/8429 

@louwie17 do you remember why?

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/DataStore.php:
```diff
@@ -13,7 +13,46 @@
+ * We use Report DataStores to separate DB data retrieval logic from the REST API controllers.
+ *
+ * Handles caching, data normalization, intervals-related methods, and other common functionality.
+ * So, in your custom report DataStore class that extend this class
```
Minor typo checked by chatgpt: should be `extends`.

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/GenericController.php:
```diff
@@ -7,10 +7,47 @@
+ * (To change the Query class or data store name, override `construct_query` method).
+ *
+ * To use it, please register your controller in the `woocommerce_admin_rest_controllers`
+ * filter and initialize the `Rest_API` class in your plugin.
```
I was thinking that if I saw this for the first time, I might assume `Rest_API` is something specific to WooCommerce that I need to initialise.

ianlin reviewed (commented):
Thanks for adding code docs. I haven't tested generating the code reference locally as I ran into a problem when building woocommerce.zip:

```
> wp --allow-root i18n make-pot . --include="woocommerce.php,client,i18n,includes,lib,packages,patterns,src,templates" --slug=woocommerce '--skip-audit'
Plugin file detected.
PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted (tried to allocate 20480 bytes) in /Users/ianlin/dev/automattic/woocommerce/plugins/woocommerce/bin/composer/wp/vendor/mck89/peast/lib/Peast/Syntax/CommentsRegistry.php on line 172
```

Probably my local env issue, but weird thing is it failed when tried to allocate memory that is a lot lower than the allowed memory. Anyway I will keep trying it tomorrow.

mikkamp commented:
> Probably my local env issue, but weird thing is it failed when tried to allocate memory that is a lot lower than the allowed memory. Anyway I will keep trying it tomorrow.

"Allowed memory size of 536870912 bytes exhausted (tried to allocate 20480 bytes)", means it tried to allocate 20k on top of the already exhausted memory.

In [other places we specifically tell it to set a higher memory limit](https://github.com/woocommerce/google-listings-and-ads/blob/2.8.2/package.json#L102) than the one you have.

tomalec committed:
Address review comments

https://github.com/woocommerce/woocommerce/pull/50746#pullrequestreview-2245593017

tomalec commented:
Thanks, @ianlin, for the review. I addressed all your comments.

> Thanks for adding code docs. I haven't tested generating the code reference locally as I ran into a problem when building woocommerce.zip:


You can also run the `code-reference` build for a released package (`./deploy.sh -v --build-only -s 9.1.4`), then replace the files in the `woocommerce` sub-directory

tomalec committed:
Make GenericController get data from the store directly. (#50749)

This allows the creation of new analytics reports without the need to create the `Query` layer. So, extension developers could use `Controller` to handle default query args in its schema and get data directly from the data store.

Co-authored-by: Mik <mikkamp@users.noreply.github.com>

tomalec committed:
Merge branch 'dev/cleanup-analytics' into dev/analytics-code-docs

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/FilteredGetDataTrait.php:
```diff
@@ -0,0 +1,58 @@
+ *
+ * @see DataStore
+ */
+trait FilteredGetDataTrait {
```
I don’t see this trait being used anywhere in the PR. Is it intended for an upcoming PR, or was it used in a different PR and is showing here due to merging?

tomalec replied:
The intent is to use in for the new classes and for 3PDs.

The existing controllers will still use `Query` classes for backward compatibility

tomalec replied:
See the discussions at https://github.com/woocommerce/woocommerce/pull/50749 as it was reviewed there.

Sorry, for merging it before review of this one is finished :(

ianlin commented on plugins/woocommerce/src/Admin/API/Reports/DataStore.php:
```diff
@@ -48,9 +48,16 @@
  *
  * Please use the `woocommerce_data_stores` filter to add your custom data store to the list of available ones.
- * Then, your store could be accessed by Query classes ({@see GenericQuery::get_data() GenericQuery::get_data()})
+ * Then, your store could be accessed by Controller classes ({@see GenericController::get_datastore_data() GenericController::get_datastore_data()})
  * or using {@link \WC_Data_Store::load() \WC_Data_Store::load()}.
```
Checking the generated report and I found the `@see` and `@link` here were not being transformed to the actual link. Would this related to the https://github.com/woocommerce/woocommerce/pull/50748#discussion_r1723766187 you added that we might need to fix the code reference?

<img width="957" alt="Screenshot 2024-08-21 at 12 18 51" src="https://github.com/user-attachments/assets/b1069cb1-979b-4c44-a710-a324c013bb83">

tomalec replied:
yes, exactly. I didn't manage the code reference to work with them. I think we can fix that later.

ianlin reviewed (commented):
@tomalec Thanks for the extra changes, also thanks @mikkamp for the explanation and tip about exhausted memory.

I left some questions below that we can have further discussion.

ianlin reviewed (approved):
Thanks for answering my questions. It's good to go, LGTM.
